### DISCUSSION TITLE
3.1.2 gives client error on ESM but not when loading from CDN

### DISCUSSION THREAD

**Discussion by Leitet** [socketio/socket.io#3828] [Q&A] [answered]
We have run into an issue that feels like a bug.

When upgrading from 3.1.1 to 3.1.2 (npm-package) we get a client error:
![image](https://user-images.githubusercontent.com/186863/110486498-fe2e0c80-80ec-11eb-9cdd-662398b3ea61.png)

This happens when loading the client module using ESM imports `import 'socket.io.js'`.
Loading the same 3.1.2 using a script tag (from /socket.io/socket.io.js or CDN) does not produce this error. 
Loading 3.1.1 with ESM import does not produce the error.

_(We are using Socket.io in a university course and the students started reporting problems last week)_

**Comment by Leitet** [comment, marked as answer]
Changing from 'window' to 'this' (this is undefined) is probably what introduced the error:
https://github.com/socketio/socket.io/blob/225ade062a13030164f89356b0a41f28203c3458/client-dist/socket.io.js#L15

_Loading a script using ESM will not set "this" to "window". This is undefined in a ESM-module. Loading the script using "script-tags" will set this to window though._

**Reply by Leitet** [reply]
It should probably be: "self"

**Reply by darrachequesne** [reply]
I could indeed reproduce, thanks. We will publish a fix as soon as possible.

**Reply by darrachequesne** [reply]
This should be fixed in `socket.io-client@3.1.3` and `socket.io-client@4.0.0`, could you please check?

**Comment by Leitet** [comment]
So after som research, it boils down to globalObject in webpack.config:
```
  output: {
    filename: "socket.io.js",
    library: "io",
    libraryTarget: "umd",
    globalObject: "this",
  },

``` 
- `self` is default and should work in every "browser case", workers, ESM and old school clientJS.
- `window` would work in ESM and osc-JS but not in workers or Node.js.
- `this` will work in osc-JS, workers and Node.js but not using ESM in client.

**Reply by Leitet** [reply]
Removing or changing to 'self' should remove the problem.
